IdentityCore/src/util/MSIDKeyOperationUtil.m

```diff
@@ -0,0 +1,130 @@
+//
+//  MSIDKeyOperationUtil.m
+//  IdentityCore
+//
+//  Created by Ameya Patil on 9/24/21.
+//  Copyright © 2021 Microsoft. All rights reserved.
+//
+
+#import "MSIDKeyOperationUtil.h"
+#import "MSIDJwtAlgorithm.h"
+#import "NSData+JWT.h"
+#import "NSData+MSIDExtensions.h"
+
+@implementation MSIDKeyOperationUtil
+
++ (MSIDKeyOperationUtil *)sharedInstance
+{
+    static dispatch_once_t once;
+    static MSIDKeyOperationUtil *s_keyOperationUtil = nil;
+    dispatch_once(&once, ^{
+        s_keyOperationUtil = [[MSIDKeyOperationUtil alloc] init];
+    });
+    return s_keyOperationUtil;
+}
+
+- (BOOL)isKeyFromSecureEnclave:(SecKeyRef)key
+{
+    if (key)
+    {
+        NSDictionary *attributes = (NSDictionary *)CFBridgingRelease(SecKeyCopyAttributes(key));
+        NSString *attrTokenId = (NSString *)[attributes valueForKey:(NSString *)CFBridgingRelease(kSecAttrTokenID)];
+        NSString *secureEnclaveId = (NSString *)(CFBridgingRelease(kSecAttrTokenIDSecureEnclave));  // id = com.apple.settoken for key in secure enclave
+        return [secureEnclaveId isEqualToString:attrTokenId];
+    }
+    return NO;
+}
+
+- (BOOL)isOperationSupportedByKey:(SecKeyOperationType)operation algorithm:(SecKeyAlgorithm)algorithm key:(SecKeyRef)key context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing *)error
+{
+    if (key == NULL)
+    {
+        [self generateErrorWithMessage:@"Key passed in to check operation is not defined." underlyingError:nil context:context error:error];
+        return NO;
+    }
+    BOOL isSupported = SecKeyIsAlgorithmSupported(key, operation, algorithm);
+    if (!isSupported)
+    {
+        NSString *errorMessage = [NSString stringWithFormat:@"Key does not support %ld with %@", (long)operation, algorithm];
+        [self generateErrorWithMessage:errorMessage underlyingError:nil context:context error:error];
+    }
+    return isSupported;
+}
+
+- (MSIDJwtAlgorithm)getJwtAlgorithmForKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key to determine signing algorithm" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    NSError *ecdsaAlgError;
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureDigestX962SHA256 key:privateKey context:context error:&ecdsaAlgError])
+    {
+        return MSID_JWT_ALG_ES256;
+    }
+
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        return MSID_JWT_ALG_RS256;
```

Comment:
Looks like this doesn't include the HMAC Algorithm option. Based on this, it's good to keep it as a separate type to avoid confusion.

Reply:
Removed the HS256 enum. This class will support signing with private key of an asymmetric key pair. If we want to add support for signing using session key in this class, we can make that change in broker later.